plan/build_key_info.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/mysql"
+)
+
+func (p *Aggregation) buildKeyInfo() {
+	p.baseLogicalPlan.buildKeyInfo()
+	// Aggregation's schema is created from aggFunc.
+	// And the ith column in schema correspond to the ith function in aggFunc.
+	// So we create a temp schema from aggFunc to build key information.
+	schema := expression.NewSchema(make([]*expression.Column, 0, p.schema.Len()))
+	for _, fun := range p.AggFuncs {
+		if col, isCol := fun.GetArgs()[0].(*expression.Column); isCol && fun.GetName() == ast.AggFuncFirstRow {
+			schema.Append(col)
+		} else {
+			// If the arg is not a column, we add a column to occupy the position.
+			schema.Append(&expression.Column{
+				FromID:   "",
+				Position: -1})
+		}
+	}
+	for _, key := range p.GetChildren()[0].GetSchema().Keys {
+		ok := true
+		newKey := make([]*expression.Column, 0, len(key))
+		for _, keyCol := range key {
+			pos := schema.GetColumnIndex(keyCol)
```

Comment:
This method can be used when dealing with p.GroupbyCols because this column can be found in child's schema.Columns. However, a key may contain multiple columns, so there will be a enumerating process which i think will cost too much.